File: src/gwaspeek/io.py
from __future__ import annotations

import json
from functools import lru_cache
from importlib import resources
from typing import Dict, List, Mapping, Optional, Sequence, Tuple

import pandas as pd


def _normalize_sep(sep: str) -> str:
    # Shell usage sometimes passes escaped separators literally (e.g. "\\t").
    escape_map = {
        r"\t": "\t",
        r"\n": "\n",
        r"\r": "\r",
    }
    return escape_map.get(sep, sep)


@lru_cache(maxsize=1)
def _column_alias_table() -> Mapping[str, object]:
    path = resources.files("gwaspeek").joinpath("data/column_aliases_from_formatbook.json")
    return json.loads(path.read_text(encoding="utf-8"))


def _header_lower_index(headers: Sequence[str]) -> Dict[str, str]:
    """First occurrence wins for case-insensitive lookup."""
    out: Dict[str, str] = {}
    for h in headers:
        key = str(h).strip().lower()
        if key not in out:
            out[key] = str(h)
    return out


def _resolve_explicit(col: str, headers: Sequence[str], lower_index: Dict[str, str]) -> str:
    if col in headers:
        return col
    key = col.strip().lower()
    if key in lower_index:
        return lower_index[key]
    preview = ", ".join(repr(c) for c in list(headers)[:12])
    raise ValueError(f"Column {col!r} not found in input header (first columns: {preview}…)")


def _pick_alias(headers_set: set[str], lower_index: Dict[str, str], aliases: Sequence[str]) -> Optional[str]:
    """Prefer longer alias names first to avoid overly generic short tokens."""
    ranked = sorted(aliases, key=lambda a: (-len(str(a)), str(a).lower()))
    for alias in ranked:
        a = str(alias)
        if a in headers_set:
            return a
        key = a.strip().lower()
        if key in lower_index:
            return lower_index[key]
    return None


def detect_sumstat_columns(
    headers: Sequence[str],
    *,
    chrom_col: Optional[str] = None,
    pos_col: Optional[str] = None,
    p_col: Optional[str] = None,
    mlog10p_col: Optional[str] = None,
) -> Tuple[str, str, Optional[str], Optional[str]]:
    """
    Map file column names to roles CHR, POS, and either P or MLOG10P.

    Returns (chrom_src, pos_src, p_src, mlog10p_src) where exactly one of
    (p_src, mlog10p_src) is not None.
    """
    headers_list = [str(h) for h in headers]
    headers_set = set(headers_list)
    lower_index = _header_lower_index(headers_list)
    spec = _column_alias_table()
    raw_aliases = spec.get("aliases", {})
    aliases_chr = list(raw_aliases.get("CHR", ["CHR"]))
    aliases_pos = list(raw_aliases.get("POS", ["POS"]))
    aliases_p = list(raw_aliases.get("P", ["P"]))
    aliases_m = list(raw_aliases.get("MLOG10P", ["MLOG10P"]))

    if chrom_col is not None:
        chrom_src = _resolve_explicit(chrom_col, headers_list, lower_index)
    else:
        chrom_src = _pick_alias(headers_set, lower_index, aliases_chr)
        if chrom_src is None:
            raise ValueError("Could not auto-detect chromosome column (CHR). Set --chr.")

    if pos_col is not None:
        pos_src = _resolve_explicit(pos_col, headers_list, lower_index)
    else:
        pos_src = _pick_alias(headers_set, lower_index, aliases_pos)
        if pos_src is None:
            raise ValueError("Could not auto-detect position column (POS). Set --pos.")

    p_src: Optional[str] = None
    mlog_src: Optional[str] = None

    if p_col is not None:
        p_src = _resolve_explicit(p_col, headers_list, lower_index)
    if mlog10p_col is not None:
        mlog_src = _resolve_explicit(mlog10p_col, headers_list, lower_index)

    if p_src is None and mlog_src is None:
        p_src = _pick_alias(headers_set, lower_index, aliases_p)
        if p_src is None:
            mlog_src = _pick_alias(headers_set, lower_index, aliases_m)
        if p_src is None and mlog_src is None:
            raise ValueError(
                "Could not auto-detect P-value or -log10(P) column. "
                "Set --p and/or --mlog10p."
            )
    elif p_src is not None and mlog_src is not None:
        raise ValueError("Specify only one of --p or --mlog10p, not both.")

    return chrom_src, pos_src, p_src, mlog_src


def load_sumstats(
    path: str,
    sep: str = "\t",
    chrom_col: Optional[str] = None,
    pos_col: Optional[str] = None,
    p_col: Optional[str] = None,
    mlog10p_col: Optional[str] = None,
) -> pd.DataFrame:
    """Load GWAS summary statistics and normalize key column names to CHR, POS, and P or MLOG10P."""
    normalized_sep = _normalize_sep(sep)
    header_df = pd.read_csv(path, sep=normalized_sep, nrows=0, low_memory=False)
    headers = list(header_df.columns)

    chrom_src, pos_src, p_src, mlog_src = detect_sumstat_columns(
        headers,
        chrom_col=chrom_col,
        pos_col=pos_col,
        p_col=p_col,
        mlog10p_col=mlog10p_col,
    )

    usecols: List[str] = [chrom_src, pos_src]
    if p_src is not None:
        usecols.append(p_src)
    else:
        if mlog_src is None:
            raise ValueError("Internal error: MLOG10P column missing after detection.")
        usecols.append(mlog_src)

    required_set = set(usecols)
    df = pd.read_csv(
        path,
        sep=normalized_sep,
        low_memory=False,
        usecols=lambda c: c in required_set,
    )

    rename_map: Dict[str, str] = {
        chrom_src: "CHR",
        pos_src: "POS",
    }
    if p_src is not None:
        rename_map[p_src] = "P"
    else:
        rename_map[mlog_src] = "MLOG10P"

    return df.rename(columns=rename_map)

File: rust/gwaspeek/src/io.rs
use flate2::read::GzDecoder;
use serde::Deserialize;
use std::collections::{HashMap, HashSet};
use std::fs::File;
use std::io::{BufReader, Read};
use std::path::Path;

use crate::preprocess::RawSumstats;

fn is_gzip_path(path: &Path) -> bool {
    path.to_string_lossy()
        .as_ref()
        .to_ascii_lowercase()
        .ends_with(".gz")
}

const ALIAS_JSON: &str = include_str!(concat!(
    env!("CARGO_MANIFEST_DIR"),
    "/../../src/gwaspeek/data/column_aliases_from_formatbook.json"
));

pub fn repo_data_dir() -> std::path::PathBuf {
    std::path::PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("../../src/gwaspeek/data")
}

pub fn normalize_sep(sep: &str) -> String {
    match sep {
        r"\t" => "\t".to_string(),
        r"\n" => "\n".to_string(),
        r"\r" => "\r".to_string(),
        _ => sep.to_string(),
    }
}

#[derive(Debug, Deserialize)]
struct AliasFile {
    aliases: HashMap<String, Vec<String>>,
}

fn column_alias_table() -> &'static AliasFile {
    static TABLE: std::sync::OnceLock<AliasFile> = std::sync::OnceLock::new();
    TABLE.get_or_init(|| serde_json::from_str(ALIAS_JSON).expect("valid column alias JSON"))
}

fn header_lower_index(headers: &[String]) -> HashMap<String, String> {
    let mut out = HashMap::new();
    for h in headers {
        let key = h.trim().to_lowercase();
        out.entry(key).or_insert_with(|| h.clone());
    }
    out
}

fn resolve_explicit(
    col: &str,
    headers: &[String],
    lower_index: &HashMap<String, String>,
) -> Result<String, String> {
    if headers.iter().any(|h| h == col) {
        return Ok(col.to_string());
    }
    let key = col.trim().to_lowercase();
    if let Some(found) = lower_index.get(&key) {
        return Ok(found.clone());
    }
    let preview: Vec<String> = headers.iter().take(12).map(|c| format!("{c:?}")).collect();
    Err(format!(
        "Column {col:?} not found in input header (first columns: {}…)",
        preview.join(", ")
    ))
}

fn pick_alias(
    headers_set: &HashSet<String>,
    lower_index: &HashMap<String, String>,
    aliases: &[String],
) -> Option<String> {
    let mut ranked: Vec<&String> = aliases.iter().collect();
    ranked.sort_by(|a, b| {
        let la = a.len();
        let lb = b.len();
        lb.cmp(&la).then_with(|| a.to_lowercase().cmp(&b.to_lowercase()))
    });
    for alias in ranked {
        let a = alias.as_str();
        if headers_set.contains(a) {
            return Some(a.to_string());
        }
        let key = a.trim().to_lowercase();
        if let Some(h) = lower_index.get(&key) {
            return Some(h.clone());
        }
    }
    None
}

/// Returns (chrom_src, pos_src, p_src, mlog10p_src) where exactly one of (p_src, mlog10p_src) is Some.
pub fn detect_sumstat_columns(
    headers: &[String],
    chrom_col: Option<&str>,
    pos_col: Option<&str>,
    p_col: Option<&str>,
    mlog10p_col: Option<&str>,
) -> Result<(String, String, Option<String>, Option<String>), String> {
    let spec = column_alias_table();
    let raw_aliases = &spec.aliases;
    let aliases_chr: Vec<String> = raw_aliases
        .get("CHR")
        .cloned()
        .unwrap_or_else(|| vec!["CHR".to_string()]);
    let aliases_pos: Vec<String> = raw_aliases
        .get("POS")
        .cloned()
        .unwrap_or_else(|| vec!["POS".to_string()]);
    let aliases_p: Vec<String> = raw_aliases
        .get("P")
        .cloned()
        .unwrap_or_else(|| vec!["P".to_string()]);
    let aliases_m: Vec<String> = raw_aliases
        .get("MLOG10P")
        .cloned()
        .unwrap_or_else(|| vec!["MLOG10P".to_string()]);

    let headers_list: Vec<String> = headers.iter().cloned().collect();
    let headers_set: HashSet<String> = headers_list.iter().cloned().collect();
    let lower_index = header_lower_index(&headers_list);

    let chrom_src = if let Some(c) = chrom_col {
        resolve_explicit(c, &headers_list, &lower_index)?
    } else {
        pick_alias(&headers_set, &lower_index, &aliases_chr)
            .ok_or_else(|| "Could not auto-detect chromosome column (CHR). Set --chr.".to_string())?
    };

    let pos_src = if let Some(c) = pos_col {
        resolve_explicit(c, &headers_list, &lower_index)?
    } else {
        pick_alias(&headers_set, &lower_index, &aliases_pos)
            .ok_or_else(|| "Could not auto-detect position column (POS). Set --pos.".to_string())?
    };

    let mut p_src: Option<String> = None;
    let mut mlog_src: Option<String> = None;
    if let Some(c) = p_col {
        p_src = Some(resolve_explicit(c, &headers_list, &lower_index)?);
    }
    if let Some(c) = mlog10p_col {
        mlog_src = Some(resolve_explicit(c, &headers_list, &lower_index)?);
    }

    if p_src.is_none() && mlog_src.is_none() {
        p_src = pick_alias(&headers_set, &lower_index, &aliases_p);
        if p_src.is_none() {
            mlog_src = pick_alias(&headers_set, &lower_index, &aliases_m);
        }
        if p_src.is_none() && mlog_src.is_none() {
            return Err(
                "Could not auto-detect P-value or -log10(P) column. Set --p and/or --mlog10p."
                    .to_string(),
            );
        }
    } else if p_src.is_some() && mlog_src.is_some() {
        return Err("Specify only one of --p or --mlog10p, not both.".to_string());
    }

    Ok((chrom_src, pos_src, p_src, mlog_src))
}

pub fn load_sumstats(
    path: &Path,
    sep: &str,
    chrom_col: Option<&str>,
    pos_col: Option<&str>,
    p_col: Option<&str>,
    mlog10p_col: Option<&str>,
) -> Result<RawSumstats, String> {
    let normalized_sep = normalize_sep(sep);
    let delim = if normalized_sep.len() == 1 {
        normalized_sep.chars().next().unwrap()
    } else {
        return Err("Multi-character separators are not supported.".to_string());
    };

    let f = File::open(path).map_err(|e| e.to_string())?;
    let reader: Box<dyn Read> = if is_gzip_path(path) {
        Box::new(BufReader::new(GzDecoder::new(f)))
    } else {
        Box::new(BufReader::new(f))
    };
    let mut rdr = csv::ReaderBuilder::new()
        .delimiter(delim as u8)
        .has_headers(true)
        .flexible(true)
        .from_reader(reader);

    let headers: Vec<String> = rdr
        .headers()
        .map_err(|e| e.to_string())?
        .iter()
        .map(|s| s.to_string())
        .collect();

    let (chrom_src, pos_src, p_src, mlog_src) =
        detect_sumstat_columns(&headers, chrom_col, pos_col, p_col, mlog10p_col)?;

    let chrom_i = headers
        .iter()
        .position(|h| h == &chrom_src)
        .ok_or_else(|| "chrom column missing".to_string())?;
    let pos_i = headers
        .iter()
        .position(|h| h == &pos_src)
        .ok_or_else(|| "pos column missing".to_string())?;
    let stat_i = if let Some(ref p) = p_src {
        headers
            .iter()
            .position(|h| h == p)
            .ok_or_else(|| "p column missing".to_string())?
    } else {
        let m = mlog_src.as_ref().unwrap();
        headers
            .iter()
            .position(|h| h == m)
            .ok_or_else(|| "mlog10p column missing".to_string())?
    };

    let mode_p = p_src.is_some();
    let mut chrom = Vec::new();
    let mut pos = Vec::new();
    let mut p_vals = Vec::new();
    let mut mlog_vals = Vec::new();

    for rec in rdr.records() {
        let rec = rec.map_err(|e| e.to_string())?;
        let c = rec.get(chrom_i).unwrap_or("").to_string();
        let p = rec.get(pos_i).unwrap_or("").to_string();
        let s = rec.get(stat_i).unwrap_or("").to_string();
        chrom.push(c);
        pos.push(p);
        if mode_p {
            p_vals.push(s);
            mlog_vals.push(String::new());
        } else {
            p_vals.push(String::new());
            mlog_vals.push(s);
        }
    }

    Ok(RawSumstats {
        chrom_tokens: chrom,
        pos_tokens: pos,
        p_tokens: p_vals,
        mlog10p_tokens: mlog_vals,
        has_p_column: mode_p,
    })
}

// In-memory variant (used by tests); file-based loader is `load_sumstats`.
pub fn load_sumstats_from_reader<R: std::io::Read>(
    r: R,
    sep: &str,
    chrom_col: Option<&str>,
    pos_col: Option<&str>,
    p_col: Option<&str>,
    mlog10p_col: Option<&str>,
) -> Result<RawSumstats, String> {
    let normalized_sep = normalize_sep(sep);
    let delim = if normalized_sep.len() == 1 {
        normalized_sep.chars().next().unwrap()
    } else {
        return Err("Multi-character separators are not supported.".to_string());
    };

    let mut rdr = csv::ReaderBuilder::new()
        .delimiter(delim as u8)
        .has_headers(true)
        .flexible(true)
        .from_reader(r);

    let headers: Vec<String> = rdr
        .headers()
        .map_err(|e| e.to_string())?
        .iter()
        .map(|s| s.to_string())
        .collect();

    let (chrom_src, pos_src, p_src, mlog_src) =
        detect_sumstat_columns(&headers, chrom_col, pos_col, p_col, mlog10p_col)?;

    let chrom_i = headers.iter().position(|h| h == &chrom_src).unwrap();
    let pos_i = headers.iter().position(|h| h == &pos_src).unwrap();
    let stat_i = if let Some(ref p) = p_src {
        headers.iter().position(|h| h == p).unwrap()
    } else {
        headers.iter().position(|h| h == mlog_src.as_ref().unwrap()).unwrap()
    };

    let mode_p = p_src.is_some();
    let mut chrom = Vec::new();
    let mut pos = Vec::new();
    let mut p_vals = Vec::new();
    let mut mlog_vals = Vec::new();

    for rec in rdr.records() {
        let rec = rec.map_err(|e| e.to_string())?;
        chrom.push(rec.get(chrom_i).unwrap_or("").to_string());
        pos.push(rec.get(pos_i).unwrap_or("").to_string());
        let s = rec.get(stat_i).unwrap_or("").to_string();
        if mode_p {
            p_vals.push(s);
            mlog_vals.push(String::new());
        } else {
            p_vals.push(String::new());
            mlog_vals.push(s);
        }
    }

    Ok(RawSumstats {
        chrom_tokens: chrom,
        pos_tokens: pos,
        p_tokens: p_vals,
        mlog10p_tokens: mlog_vals,
        has_p_column: mode_p,
    })
}

#[cfg(test)]
mod tests {
    use super::*;
    use flate2::write::GzEncoder;
    use flate2::Compression;
    use std::io::Write;
    use std::path::PathBuf;

    fn fixture(name: &str) -> PathBuf {
        PathBuf::from(env!("CARGO_MANIFEST_DIR"))
            .join("../../tests/fixtures")
            .join(name)
    }

    #[test]
    fn detect_columns_default_fixture_headers() {
        let p = fixture("sumstats_small.tsv");
        let raw = load_sumstats(&p, "\t", None, None, None, None).unwrap();
        assert!(raw.has_p_column);
        assert_eq!(raw.chrom_tokens.len(), 7);
    }

    #[test]
    fn load_escaped_tab_sep() {
        let p = fixture("sumstats_small.tsv");
        let raw = load_sumstats(&p, r"\t", None, None, None, None).unwrap();
        assert_eq!(raw.chrom_tokens.len(), 7);
    }

    #[test]
    fn load_gzipped_fixture_matches_plain() {
        let p = fixture("sumstats_small.tsv");
        let plain = load_sumstats(&p, "\t", None, None, None, None).unwrap();

        let data = std::fs::read(&p).unwrap();
        let mut gz_path = std::env::temp_dir();
        gz_path.push(format!(
            "gwaspeek_gz_test_{}.tsv.gz",
            std::process::id()
        ));
        {
            let f = File::create(&gz_path).unwrap();
            let mut enc = GzEncoder::new(f, Compression::default());
            enc.write_all(&data).unwrap();
            enc.finish().unwrap();
        }
        let gz = load_sumstats(&gz_path, "\t", None, None, None, None).unwrap();
        let _ = std::fs::remove_file(&gz_path);

        assert_eq!(gz.chrom_tokens, plain.chrom_tokens);
        assert_eq!(gz.pos_tokens, plain.pos_tokens);
        assert_eq!(gz.p_tokens, plain.p_tokens);
        assert_eq!(gz.mlog10p_tokens, plain.mlog10p_tokens);
        assert_eq!(gz.has_p_column, plain.has_p_column);
    }
}
